ARCHITECTURE.md

```diff
@@ -0,0 +1,132 @@
+
+  - manpages from releases of the git project, imported and formatted
+    via asciidoctor; see the `lib/tasks/index.rake` task.
+
+
+## Heroku
+
+The app itself is served by Heroku. The app name is `git-scm` (so you
+can visit it directly as https://git-scm.herokuapp.com). The site is
+owned by the git-scm.com team. If you want to be involved in managing
+uptime/deploys/etc, you'll need a Heroku account and request to be added
+to that team. The git-scm team receives credits from Heroku so that the
+hosting is free.
+
+We use a few Heroku add-ons:
+
+  - Bonsai elasticsearch (see below)
+
+  - Heroku Postgres as the database
+
+  - Heroku Redis for rails caching
+
+  - Heroku scheduler for cron jobs
+
+The nightly scheduled jobs are:
+
+  - `rake downloads` (pick up newly released git versions)
+
+  - `rake preindex` (pull in and format manpages for released git
+    versions)
+
+  - `rake remote_genbook2` (pull in and format progit2 book content,
+    including translations)
+
+It should be safe to run any of those jobs more frequently. E.g., if you
+know there's a new Git release out, then:
+
+    heroku run rake preindex
+    heroku run rake downloads
+
+will get it on the site without waiting for the nightly run.
+
+Merges to the `master` branch on GitHub auto-deploy to Heroku, so unless
+you're doing something tricky you generally shouldn't need to manually
+deploy.
+
+Note that some of the formatting of manpages and book content happens
+when they are imported by the rake tasks. So after fixing some
+formatting and deploying, the rake jobs may need to be re-run with a
+special flag to re-import (see the individual tasks for details).
+
+
+## Cloudflare
+
+We get enough requests that it's easy to overwhelm the single Heroku
+dyno. So we have Cloudflare sitting in front of it, aggressively caching
+everything. That also should make the site faster to serve to regions
+far away from Heroku's servers.
+
+The Cloudflare setup is mostly pretty simple:
+
+ - they serve DNS for the whole domain (that's where they insert the CDN
+   magic)
+
+ - Cloudflare provides `https://` support to the user. Obviously the
+   site is totally open and doesn't have any sensitive data, so this is
+   really more about integrity. The certificate is generated by
+   Cloudflare (and requires SNI on the browser side).
+
+ - the Cloudflare connection to Heroku is passed over TLS; they provide an
+   "internal" certificate that we ask Heroku to use, so the connection
+   is secured between the two (again, mostly for integrity)
+
+ - the most exotic config is that we use "page rules" to mark the whole
+   site to be cached aggressively, regardless of any caching headers
+   sent from Heroku. This is a bit of a hack, but there's very little on
+   the site that can't be cached (which is perhaps a sign that the rails
+   setup needs to be tweaked to send more reasonable caching headers,
+   but this has been simple and effective so far).
+
```

Comment:
should we mention the /downloads/win (or something similar) exception we did some time ago due to the windows version auto detection?

Reply:
Thanks, I meant to mention that but forgot. Fixed.